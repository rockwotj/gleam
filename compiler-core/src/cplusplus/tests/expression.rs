use crate::assert_cpp;

#[test]
fn boolean_operators() {
    assert_cpp!(
        r#"
pub fn or_op(x, y) {
  x < 5 || y < 10
}
pub fn and_op(x, y) {
  x < 5 && y < 10
}
pub fn both_ops(x, y) {
  { x < 5 || y < 10 } && { x > 5 || y > 10 }
}"#
    );
}

#[test]
fn variables() {
    assert_cpp!(
        r#"
pub fn var_test(a: Float, b: Float, c: Float) -> Float {
  let x = a;
  let y = b;
  let z = c;
  let q = a +. x +. y +. z;
  let v = x +. b +. c;
  v +. q +. z +. y +. x
}
    "#
    )
}

#[test]
fn record_access() {
    assert_cpp!(
        r#"
pub type Person {
  Person(name: String, age: Int)
}

pub fn next_age(person: Person) -> Int {
  person.age + 1
}
    "#
    )
}

#[test]
fn record_construction() {
    assert_cpp!(
        r#"
pub type Person {
  Person(name: String, age: Int)
}

pub fn increment(n: Int) -> Int {
  n + 1
}

pub fn birthday(person: Person) -> Person {
  let next_age = increment(person.age);
  Person(person.name, next_age)
}
    "#
    )
}

#[test]
fn record_construction_labels() {
    assert_cpp!(
        r#"
pub type Person {
  Person(name: String, age: Int)
}

pub fn increment(n: Int) -> Int {
  n + 1
}

pub fn birthday(person: Person) -> Person {
  let next_age = increment(person.age);
  Person(age: next_age, name: person.name)
}
    "#
    )
}
