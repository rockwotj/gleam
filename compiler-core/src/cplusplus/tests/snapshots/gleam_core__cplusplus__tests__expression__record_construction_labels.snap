---
source: compiler-core/src/cplusplus/tests/expression.rs
expression: "\npub type Person {\n  Person(name: String, age: Int)\n}\n\npub fn increment(n: Int) -> Int {\n  n + 1\n}\n\npub fn birthday(person: Person) -> Person {\n  let next_age = increment(person.age);\n  Person(age: next_age, name: person.name)\n}\n    "
---
#ifndef MY_PACKAGE_MY_MODULE_H_
#define MY_PACKAGE_MY_MODULE_H_

#include <gleam.h>

namespace my {
namespace module {

class Person;

int64_t increment(int64_t n);

gleam::Ref<::my::module::Person> birthday(gleam::Ref<::my::module::Person> person);

class Person {
public:
  Person(gleam::Ref<gleam::String> name, int64_t age) : m$name(name), m$age(age) {}
  gleam::Ref<gleam::String> name() const { return m$name; }
  int64_t age() const { return m$age; }
private:
  gleam::Ref<gleam::String> m$name;
  int64_t m$age;
};

} // namespace my
} // namespace module

#endif // MY_PACKAGE_MY_MODULE_H_

---
#include "module.h"

namespace my {
namespace module {

int64_t increment(int64_t n) {
  return n + 1;
};

gleam::Ref<::my::module::Person> birthday(gleam::Ref<::my::module::Person> person) {
  int64_t next_age = ::my::module::_private::increment(person->age());
  return ::my::module::_private::Person(person->name(),
  next_age);
};

} // namespace my
} // namespace module

